### ISSUE TITLE
Using ibm_db with Electron

### ISSUE THREAD

**Issue by alexeycv** [ibmdb/node-ibm_db#336]
Hi!

I'm trying to use ibm_db module with Electron application on Windows.

I use this code to connect to my db2 server:
```
import * as db from "ibm_db";
let connectionString = "DATABASE=[db];HOSTNAME=[host];UID=[username];PWD=[password];;PORT=50000;PROTOCOL=TCPIP";
db.open(connectionString, function(error, conn){
    console.log("Opened");
});
```
And have following error when db.open(...) is executed:
```
Problem signature:
  Problem Event Name:	APPCRASH
  Application Name:	FATHOM Aggregator.exe
  Application Version:	1.8.1.0
  Application Timestamp:	59cd494d
  Fault Module Name:	ODBC32.dll
  Fault Module Version:	6.3.9600.17415
  Fault Module Timestamp:	5450504d
  Exception Code:	c0000005
  Exception Offset:	000000000000a4e9
  OS Version:	6.3.9600.2.0.0.272.7
  Locale ID:	1033
  Additional Information 1:	186c
  Additional Information 2:	186c79fb46894675ea2a45ce3771a6cf
  Additional Information 3:	aec7
  Additional Information 4:	aec7afc783f01d450e5037076227c879
```
My environment:
OS Windows Server 2012 R2
NodeJS 8.8.1 (x64)
Electron 1.8.1
node-ibm_db 2.2.1

Did someone have such error or know how to fix them?

When I use clean NodeJS without Electron it run without any problem.

**Comment by rhtpandeyIN** [comment]
@alexeycv, Are you getting any error while installing ibm_db using electron?

**Comment by alexeycv** [comment]
@rhtpandeyIN I'm getting only "cli driver build fail" error, but binary driver is installed correctly.

**Comment by rhtpandeyIN** [comment]
Would you be able to provide full installation log? and if possible steps to repro this situation.

**Comment by alexeycv** [comment]
Installation lig
```
C:\Projects\Alexey\sample_app>npm install ibm_db

> ibm_db@2.2.1 install C:\Projects\Alexey\sample_app\node_modules\ibm_db
> node installer/driverInstall.js

Downloading DB2 ODBC CLI Driver from https://public.dhe.ibm.com/ibmdl/export/pub
/software/data/db2/drivers/odbc_cli/ntx64_odbc_cli.zip...

100.00% | 17935538 bytes downloaded out of 17935538 bytes.

****************************************
You are downloading a package which includes the Node.js module for IBM DB2/Info
rmix.  The module is licensed under the Apache License 2.0. The package also inc
ludes IBM ODBC and CLI Driver from IBM, which is automatically downloaded as the
 node module is installed on your system/device. The license agreement to the IB
M ODBC and CLI Driver is available in undefined   Check for additional dependenc
ies, which may come with their own license agreement(s). Your use of the compone
nts of the package and dependencies constitutes your acceptance of their respect
ive license agreements. If you do not accept the terms of any license agreement(
s), then delete the relevant component(s) from your device.
****************************************

Downloading and extraction of DB2 ODBC CLI Driver completed successfully...


C:\Projects\Alexey\sample_app\node_modules\ibm_db>if not defined npm_config_node
_gyp (node "C:\Users\Administrator\AppData\Roaming\npm\node_modules\npm\bin\node
-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" configure build  --IS_DOW
NLOADED=true --IBM_DB_HOME=$IBM_DB_HOME )  else (node "" configure build  --IS_D
OWNLOADED=true --IBM_DB_HOME=$IBM_DB_HOME )
Building the projects in this solution one at a time. To enable parallel build,
please add the "/m" switch.
  odbc.cpp
  odbc_connection.cpp
c:\projects\alexey\sample_app\node_modules\ibm_db\installer\clidriver\include\sq
lcli.h(59): warning C4005: 'FAR': macro redefinition (compiling source file ..\s
rc\odbc.cpp) [C:\Projects\Alexey\sample_app\node_modules\ibm_db\build\odbc_bindi
ngs.vcxproj]
  C:\Program Files (x86)\Windows Kits\8.1\Include\shared\minwindef.h(150): note:
 see previous definition of 'FAR' (compiling source file ..\src\odbc.cpp)
c:\projects\alexey\sample_app\node_modules\ibm_db\installer\clidriver\include\sq
lcli.h(59): warning C4005: 'FAR': macro redefinition (compiling source file ..\s
rc\odbc_connection.cpp) [C:\Projects\Alexey\sample_app\node_modules\ibm_db\build
\odbc_bindings.vcxproj]
  C:\Program Files (x86)\Windows Kits\8.1\Include\shared\minwindef.h(150): note:
 see previous definition of 'FAR' (compiling source file ..\src\odbc_connection.
cpp)
..\src\odbc.cpp(204): warning C4996: 'v8::Function::NewInstance': was declared d
eprecated [C:\Projects\Alexey\sample_app\node_modules\ibm_db\build\odbc_bindings
.vcxproj]
  C:\Users\Administrator\.node-gyp\8.8.1\include\node\v8.h(3832): note: see decl
aration of 'v8::Function::NewInstance'
..\src\odbc.cpp(247): warning C4996: 'v8::Function::NewInstance': was declared d
eprecated [C:\Projects\Alexey\sample_app\node_modules\ibm_db\build\odbc_bindings
.vcxproj]
  C:\Users\Administrator\.node-gyp\8.8.1\include\node\v8.h(3832): note: see decl
aration of 'v8::Function::NewInstance'
..\src\odbc.cpp(845): warning C4244: '=': conversion from 'int64_t' to 'SQLSMALL
INT', possible loss of data [C:\Projects\Alexey\sample_app\node_modules\ibm_db\b
uild\odbc_bindings.vcxproj]
..\src\odbc.cpp(849): warning C4244: '=': conversion from 'int64_t' to 'SQLSMALL
INT', possible loss of data [C:\Projects\Alexey\sample_app\node_modules\ibm_db\b
uild\odbc_bindings.vcxproj]
..\src\odbc.cpp(855): warning C4244: '=': conversion from 'int64_t' to 'SQLSMALL
INT', possible loss of data [C:\Projects\Alexey\sample_app\node_modules\ibm_db\b
uild\odbc_bindings.vcxproj]
..\src\odbc.cpp(863): warning C4244: '=': conversion from 'int64_t' to 'SQLINTEG
ER', possible loss of data [C:\Projects\Alexey\sample_app\node_modules\ibm_db\bu
ild\odbc_bindings.vcxproj]
..\src\odbc.cpp(969): warning C4244: '=': conversion from 'SQLINTEGER' to 'SQLSM
ALLINT', possible loss of data [C:\Projects\Alexey\sample_app\node_modules\ibm_d
b\build\odbc_bindings.vcxproj]
..\src\odbc_connection.cpp(579): warning C4996: 'v8::Function::NewInstance': was
 declared deprecated [C:\Projects\Alexey\sample_app\node_modules\ibm_db\build\od
bc_bindings.vcxproj]
  C:\Users\Administrator\.node-gyp\8.8.1\include\node\v8.h(3832): note: see decl
aration of 'v8::Function::NewInstance'
..\src\odbc.cpp(1180): warning C4244: 'argument': conversion from 'SQLINTEGER' t
o 'SQLSMALLINT', possible loss of data [C:\Projects\Alexey\sample_app\node_modul
es\ibm_db\build\odbc_bindings.vcxproj]
..\src\odbc_connection.cpp(663): warning C4996: 'v8::Function::NewInstance': was
 declared deprecated [C:\Projects\Alexey\sample_app\node_modules\ibm_db\build\od
bc_bindings.vcxproj]
..\src\odbc.cpp(1189): warning C4996: 'v8::Object::SetPrototype': was declared d
eprecatedC:\Users\Administrator\.node-gyp\8.8.1\include\node\v8.h(3832): note: s
ee declaration of 'v8::Function::NewInstance' [C:\Projects\Alexey\sample_app\nod
e_modules\ibm_db\build\odbc_bindings.vcxproj]

  C:\Users\Administrator\.node-gyp\8.8.1\include\node\v8.h(3256): note: see decl
aration of 'v8::Object::SetPrototype'
..\src\odbc_connection.cpp(906): warning C4996: 'v8::Function::NewInstance': was
 declared deprecated [C:\Projects\Alexey\sample_app\node_modules\ibm_db\build\od
bc_bindings.vcxproj]
  C:\Users\Administrator\.node-gyp\8.8.1\include\node\v8.h(3832): note: see decl
aration of 'v8::Function::NewInstance'
..\src\odbc_connection.cpp(1136): warning C4996: 'v8::Function::NewInstance': wa
s declared deprecated [C:\Projects\Alexey\sample_app\node_modules\ibm_db\build\o
dbc_bindings.vcxproj]
  C:\Users\Administrator\.node-gyp\8.8.1\include\node\v8.h(3832): note: see decl
aration of 'v8::Function::NewInstance'
  odbc_statement.cpp
  odbc_result.cpp
c:\projects\alexey\sample_app\node_modules\ibm_db\installer\clidriver\include\sq
lcli.h(59): warning C4005: 'FAR': macro redefinition (compiling source file ..\s
rc\odbc_statement.cpp) [C:\Projects\Alexey\sample_app\node_modules\ibm_db\build\
odbc_bindings.vcxproj]
  C:\Program Files (x86)\Windows Kits\8.1\Include\shared\minwindef.h(150): note:
 see previous definition of 'FAR' (compiling source file ..\src\odbc_statement.c
pp)
..\src\odbc_statement.cpp(217): warning C4996: 'v8::Function::NewInstance': was
declared deprecated [C:\Projects\Alexey\sample_app\node_modules\ibm_db\build\odb
c_bindings.vcxproj]
  C:\Users\Administrator\.node-gyp\8.8.1\include\node\v8.h(3832): note: see decl
aration of 'v8::Function::NewInstance'
..\src\odbc_statement.cpp(289): warning C4996: 'v8::Function::NewInstance': was
declared deprecated [C:\Projects\Alexey\sample_app\node_modules\ibm_db\build\odb
c_bindings.vcxproj]
  C:\Users\Administrator\.node-gyp\8.8.1\include\node\v8.h(3832): note: see decl
aration of 'v8::Function::NewInstance'
c:\projects\alexey\sample_app\node_modules\ibm_db\installer\clidriver\include\sq
lcli.h(59): warning C4005: 'FAR': macro redefinition (compiling source file ..\s
rc\odbc_result.cpp) [C:\Projects\Alexey\sample_app\node_modules\ibm_db\build\odb
c_bindings.vcxproj]
  C:\Program Files (x86)\Windows Kits\8.1\Include\shared\minwindef.h(150): note:
 see previous definition of 'FAR' (compiling source file ..\src\odbc_result.cpp)

..\src\odbc_statement.cpp(531): warning C4996: 'v8::Function::NewInstance': was
declared deprecated [C:\Projects\Alexey\sample_app\node_modules\ibm_db\build\odb
c_bindings.vcxproj]
  C:\Users\Administrator\.node-gyp\8.8.1\include\node\v8.h(3832): note: see decl
aration of 'v8::Function::NewInstance'
..\src\odbc_statement.cpp(596): warning C4996: 'v8::Function::NewInstance': was
declared deprecated [C:\Projects\Alexey\sample_app\node_modules\ibm_db\build\odb
c_bindings.vcxproj]
  C:\Users\Administrator\.node-gyp\8.8.1\include\node\v8.h(3832): note: see decl
aration of 'v8::Function::NewInstance'
..\src\odbc_result.cpp(183): warning C4996: 'v8::Value::ToInt32': was declared d
eprecated [C:\Projects\Alexey\sample_app\node_modules\ibm_db\build\odbc_bindings
.vcxproj]
  C:\Users\Administrator\.node-gyp\8.8.1\include\node\v8.h(9743): note: see decl
aration of 'v8::Value::ToInt32'
..\src\odbc_result.cpp(337): warning C4996: 'v8::Value::ToInt32': was declared d
eprecated [C:\Projects\Alexey\sample_app\node_modules\ibm_db\build\odbc_bindings
.vcxproj]
  C:\Users\Administrator\.node-gyp\8.8.1\include\node\v8.h(9743): note: see decl
aration of 'v8::Value::ToInt32'
..\src\odbc_result.cpp(435): warning C4996: 'v8::Value::ToInt32': was declared d
eprecated [C:\Projects\Alexey\sample_app\node_modules\ibm_db\build\odbc_bindings
.vcxproj]
  C:\Users\Administrator\.node-gyp\8.8.1\include\node\v8.h(9743): note: see decl
aration of 'v8::Value::ToInt32'
..\src\odbc_result.cpp(618): warning C4996: 'v8::Value::ToInt32': was declared d
eprecated [C:\Projects\Alexey\sample_app\node_modules\ibm_db\build\odbc_bindings
.vcxproj]
  C:\Users\Administrator\.node-gyp\8.8.1\include\node\v8.h(9743): note: see decl
aration of 'v8::Value::ToInt32'
  win_delay_load_hook.cc
  strptime.c
     Creating library C:\Projects\Alexey\sample_app\node_modules\ibm_db\build\Re
lease\odbc_bindings.lib and object C:\Projects\Alexey\sample_app\node_modules\ib
m_db\build\Release\odbc_bindings.exp
  Generating code
  Finished generating code
  odbc_bindings.vcxproj -> C:\Projects\Alexey\sample_app\node_modules\ibm_db\bui
ld\Release\\odbc_bindings.node
  odbc_bindings.vcxproj -> C:\Projects\Alexey\sample_app\node_modules\ibm_db\bui
ld\Release\odbc_bindings.pdb (Full PDB)


Kernel additional dependencies removed successfully!


{ Error: Command failed: msbuild /clp:Verbosity=minimal /nologo /p:Configuration
=Release;Platform=x64 C:\Projects\Alexey\sample_app\node_modules\ibm_db\build\bi
nding.sln
'msbuild' is not recognized as an internal or external command,
operable program or batch file.

    at ChildProcess.exithandler (child_process.js:272:12)
    at emitTwo (events.js:125:13)
    at ChildProcess.emit (events.js:213:7)
    at maybeClose (internal/child_process.js:927:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:211:5)
  killed: false,
  code: 1,
  signal: null,
  cmd: 'msbuild /clp:Verbosity=minimal /nologo /p:Configuration=Release;Platform
=x64 C:\\Projects\\Alexey\\sample_app\\node_modules\\ibm_db\\build\\binding.sln'
 }

msbuild build process failed!
Proceeding with Pre-compiled Binary Installation.


===================================
node-ibm_db installed successfully!
===================================
```

This code cause application to crash (in main or in render process, no matter):
``` javascript
const db require("ibm_db");
...
let connectionString = "DATABASE=[db];HOSTNAME=localhost;UID=[userid];PWD=[password];;PORT=50000;PROTOCOL=TCPIP";
db.open(connectionString, function(error, conn){
	console.log("Opened");
});
```

**Comment by rhtpandeyIN** [comment]
I think electron has some problem with pre-compiled binaries, for that we introduced cli driver build/compilation part on windows. As per the log, it seems you have Visual Studio installed but "msbuild" path is not set or may be "msbuild" doesn't exist in your system.

Please try to set "msbuild" in your environment path variable, and then install ibm_db using "cmd" with admin privileges.
I think it should work then.

**Comment by alexeycv** [comment]
I install ms build tools 14.0 and set path to msbuild executable. After that, I install ibm_db, successfully without errors (from sources). But, when I run application, I have the same error.
```
Problem signature:
  Problem Event Name:	APPCRASH
  Application Name:	electron.exe
  Application Version:	1.8.1.0
  Application Timestamp:	59cd494d
  Fault Module Name:	ODBC32.dll
  Fault Module Version:	6.3.9600.17415
  Fault Module Timestamp:	5450504d
  Exception Code:	c0000005
  Exception Offset:	000000000000a4e9
  OS Version:	6.3.9600.2.0.0.272.7
  Locale ID:	1033
  Additional Information 1:	186c
  Additional Information 2:	186c79fb46894675ea2a45ce3771a6cf
  Additional Information 3:	cc1a
  Additional Information 4:	cc1a631e7b53259a805cadcf3ff5ca23
```
when running application by command "electron ."

**Comment by LitvK** [comment]
Any changes in this situation?
Some new ideas maybe?

**Comment by bimalkjha** [comment]
@alexeycv @LitvK I think this electron issue is fixed in ibm_db@2.3.1. Please verify. Thanks.
